cpp/src/arrow/python/helpers.cc

```diff
@@ -193,6 +195,32 @@ Status UInt64FromPythonInt(PyObject* obj, uint64_t* out) {
   return Status::OK();
 }
 
+bool PyFloat_isnan(const PyObject* obj) {
+  return PyFloat_Check(obj) && std::isnan(PyFloat_AS_DOUBLE(obj));
+}
+
+bool PyDecimal_Check(const PyObject* obj) {
+  // TODO(phillipc): Is this expensive? Can we make it a global?
+  OwnedRef Decimal;
+  {
+    OwnedRef decimal;
+    Status status = ImportModule("decimal", &decimal);
+    DCHECK(status.ok()) << "Error during import of the decimal module";
+    status = ImportFromModule(decimal, "Decimal", &Decimal);
+    DCHECK(status.ok())
+        << "Error during import of the Decimal object from the decimal module";
+  }
+  const int32_t result = PyObject_IsInstance(const_cast<PyObject*>(obj), Decimal.obj());
+  DCHECK_NE(result, -1) << " error during PyObject_IsInstance check";
+  return result == 1;
+}
+
+bool PyDecimal_isnan(const PyObject* obj) {
+  OwnedRef is_nan(PyObject_CallMethod(
+      const_cast<PyObject*>(obj), const_cast<char*>("is_nan"), const_cast<char*>("()")));
+  return PyDecimal_Check(obj) && is_nan.obj() == Py_True;
```

Comment:
You should call `PyDecimal_Check` *before* calling the method. Also you should check if the method raises. And it's better to use `PyObject_IsTrue` rather than compare against `Py_True`.

Reply:
Yes, not sure how I missed that :)